### PR TITLE
fix: Make Flow Grid compatible with ASM versions older than 9

### PR DESCRIPTION
**Pull request by mshabarov**
Solution for https://github.com/vaadin/flow/issues/15625, until https://github.com/payara/Payara/issues/6127 is fixed.

### CODE REVIEW

**Review by mcollovati — APPROVED** [review]

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 1]
```diff
@@ -358,31 +372,12 @@ protected void fireSelectionEvent(
...
-
-        /**
-         * Selection model that doesn't allow selection.
-         *
-         * @see GridNoneSelectionModel
-         */
-        NONE {
-            @Override
-            protected <T> GridSelectionModel<T> createModel(Grid<T> grid) {
+            } else if (NONE.equals(this)) {
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:375]: @sonatype-lift ignoreall

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 2]
```diff
@@ -318,9 +318,34 @@ public enum SelectionMode {
...
+        /**
+         * Creates the selection model to use with this enum.
+         *
+         * @param <T>
+         *            the type of items in the grid
+         * @param grid
+         *            the grid to create the selection model for
+         * @return the selection model
+         */
+        protected <T> GridSelectionModel<T> createModel(Grid<T> grid) {
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:347]: @sonatype-lift ignoreall

**Review by MarcinVaadin — APPROVED** [review]

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 3]
```diff
@@ -318,9 +318,34 @@ public enum SelectionMode {
...
+            if (SINGLE.equals(this)) {
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:348]: Any reason not to use a `switch` here?

Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java]: Only a personal preferences. Changed to enhanced `switch`

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java [hunk 4]
```diff
@@ -382,7 +344,34 @@ protected <T> GridSelectionModel<T> createModel(Grid<T> grid) {
          *            the grid to create the selection model for
          * @return the selection model
          */
-        protected abstract <T> GridSelectionModel<T> createModel(Grid<T> grid);
+        protected <T> GridSelectionModel<T> createModel(Grid<T> grid) {
+            return switch (this) {
+            case SINGLE -> new AbstractGridSingleSelectionModel<T>(grid) {
```
Comment (knoobie) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:349]: Just thinking out load: are you sure the oldish ASM can work with this newish switch? 😬

Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/Grid.java:349]: Good catch! But believe or not, this new switch doesn't trigger any errors in Payara, application starts fine  :sweat_smile: don't ask me why :)

**Comment by mshabarov** [comment]
We need this before releasing 24.0 beta.

**Review by yuriy-fix — APPROVED** [review]

**Pull request merged**